File: src/main/java/core/procedures/math/complex/RealPart.java
package core.procedures.math.complex;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMBigComplex;

public final class RealPart extends AFn {

  public RealPart() {
    super(new FnArgsBuilder().minArgs(1).maxArgs(1).mandatoryArgsTypes(new Class[]{Number.class}));
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "real-part";
  }

  @Override
  public Number apply1(Object arg) {
    if (arg instanceof SCMBigComplex) {
      return ((SCMBigComplex)arg).getRe();
    }
    return (Number)arg;
  }
}

File: src/main/java/core/procedures/math/complex/RealPart.kt
package core.procedures.math.complex

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Complex

class RealPart : AFn<Number?, Number>(name = "real-part", isPure = true, arity = Exactly(1),
                     mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?) = when (arg) {
        is Complex -> arg.re
        else -> arg!!
    }
}
